### ISSUE TITLE
Does not recognise Vitest 2 inside `beforeAll` hook

### ISSUE THREAD

**Issue by RobinTail** [thebuilder/react-intersection-observer#689]
**Describe the bug** A clear and concise description of what the bug is.

When trying to upgrade Vitest from 1.6 to 2.0 I'm getting these warnings:

```
React Intersection Observer was not configured to handle mocking.
Outside Jest and Vitest, you might need to manually configure it by calling setupIntersectionMocking() and resetIntersectionMocking() in your test setup file.

// test-setup.js
import { resetIntersectionMocking, setupIntersectionMocking } from 'react-intersection-observer/test-utils';

beforeEach(() => {
  setupIntersectionMocking(vi.fn);
});

afterEach(() => {
  resetIntersectionMocking();
});
```

**To Reproduce** Try and recreate the issue in a **Codesandbox**:

none yet

**Expected behavior** A clear and concise description of what you expected to
happen.

Should detect Vitest 2 similar to Vitest 1.6

**Additional context** Add any other context about the problem here.

Running latest version, 9.10.3

**Comment by thebuilder** [comment]
Interesting. I'll have a look. 
Also what to see how it works with the new browser mode in Vitest.

**Comment by thebuilder** [comment]
@RobinTail are you sure it's running with `global: true` in Vitest config?

I've tried upgrading the library itself to Vitest 2, and it doesn't break anything.

**Comment by RobinTail** [comment]
> @RobinTail are you sure it's running with `global: true` in Vitest config?

absolutely! I'm surprised myself.

**Comment by RobinTail** [comment]
Maybe it's because of this settings. I will try another pool.

```jsonc
    pool: "vmThreads",
    poolOptions: {
      vmThreads: { useAtomics: true, memoryLimit: 0.5 },
    },
```

**Comment by RobinTail** [comment]
No, `pool: "forks"` does not fix this.

**Comment by thebuilder** [comment]
#690 The tests using the `test-utils` in the library are all running fine after upgrading.

**Comment by RobinTail** [comment]
I'll try to prepare a sandbox to reproduce it

**Comment by RobinTail** [comment]
Dear @thebuilder , here is the VM that reproduces the issue:

https://codesandbox.io/p/devbox/crimson-night-w64sxn

In terminal try to execute `yarn test`.

**Comment by RobinTail** [comment]
Perhaps it may be related to the some changes of the hooks execution order in Vitest 2.
Because I'm calling the method inside `beforeAll` hook.
I'm not sure what exactly they mean by "reverse order"

<img width="681" alt="Screenshot 2024-07-10 at 21 43 44" src="https://github.com/thebuilder/react-intersection-observer/assets/13189514/314fc345-e805-46ec-bdc8-9db34a705388">

**Comment by thebuilder** [comment]
Thanks a lot - I think the problem (in the example) is that you trigger `mockAllIsIntersecting` in a `beforeAll`. It's running before the `beforeAll` in `test-utils` had a chance to run.

If I change it so `mockAllIsIntersecting` is triggered inside the `test` it works:

```ts
import { mockAllIsIntersecting } from "react-intersection-observer/test-utils";

describe("Test", () => {
  test("test", () => {
    mockAllIsIntersecting(true);
    expect(true).toBeTruthy();
  });
});
```

**Comment by RobinTail** [comment]
Yes, @thebuilder , BUT :)
If I downgrade Vitest to 1.6.0, it works inside `beforeAll` without warnings.

**Comment by RobinTail** [comment]
> It's running before the `beforeAll` in `test-utils` is triggered.

I see it's actually `beforeEach` in `test-utils`:

https://github.com/thebuilder/react-intersection-observer/blob/7feb5b69ae0bcff023abc747395b764d2cd09f5c/src/test-utils.ts#L24

So maybe in Vitest 2 `beforeEach` runs after `beforeAll`

**Comment by RobinTail** [comment]
The thing is that `beforeAll` is the recommended place to establish mocking

![image](https://github.com/thebuilder/react-intersection-observer/assets/13189514/13646378-e475-4c95-9f4e-44572db4a61d)

**Comment by thebuilder** [comment]
It runs in `beforeEach` to reset the mock between tests. 

Regardless, I'm not sure why you are triggering the `mockAllIsIntersecting` before the test is running. 🤔 The intention is you trigger it after creating the Intersection Observers. It only affects active observers, so if you haven't created any, then it won't do anything.

**Comment by RobinTail** [comment]
Ok

**Comment by RobinTail** [comment]
> It runs in beforeEach to reset the mock between tests.

Reset mocking you do in `afterEach`, FYI:

https://github.com/thebuilder/react-intersection-observer/blob/7feb5b69ae0bcff023abc747395b764d2cd09f5c/src/test-utils.ts#L33-L35

**Comment by thebuilder** [comment]
[`f72e455` (#690)](https://github.com/thebuilder/react-intersection-observer/pull/690/commits/f72e4559d4ad0d1188bee31db67d17fb695b2e30)

**Comment by thebuilder** [comment]
FYI @RobinTail - I'll drop the `beforeAll` in #716 
It's simply not safe to rely on just the `beforeAll`, since global options in test runners like `mockReset` can reset the observer after a test run, preventing it from working in the following.

- Don't run  `mockAllIsIntersecting` in `beforeAll`
- Consider using a [fallback](https://github.com/thebuilder/react-intersection-observer#fallback-behavior) instead, to always have observers report `true` if they aren't mocked.

**Comment by condorheroblog** [comment]
Hi @thebuilder.

I also encountered this problem.

<img width="1159" alt="Image" src="https://github.com/user-attachments/assets/6319a40c-90f7-423b-9cee-a2cf8f1eab0d" />

I didn't turn on the global mode. I used the `mockIsIntersecting` function in the test file, which called the `warnOnMissingSetup` function. Tracing the call chain, I found that this function caused it.

```
function testLibraryUtil(): typeof vi | undefined {
  if (typeof vi !== "undefined") return vi;
  // @ts-expect-error We don't include the Jest types
  if (typeof jest !== "undefined") return jest;
  return undefined;
}
```

Because I don't have global mode enabled, this function will return undefined and print a warning message.

---

This is how I use it.

`vitest.config.ts@3.0.7`

```ts
/// <reference types="vitest/config" />

import { defineConfig } from "vite";

export default defineConfig({
  test: {
    environment: "happy-dom",
    setupFiles: ["./tests/vitest.setup.ts"],
  },
});
```

1. `vitest.setup.ts`

```ts
import "@testing-library/jest-dom/vitest";
import { cleanup } from "@testing-library/react";
import {
  resetIntersectionMocking,
  setupIntersectionMocking,
} from "react-intersection-observer/test-utils";
import { afterEach, beforeEach, vi } from "vitest";

beforeEach(() => {
  setupIntersectionMocking(vi.fn);
});

afterEach(() => {
  resetIntersectionMocking();
  cleanup();
});
```

2. mockIsIntersecting

```
import { act, render, screen } from "@testing-library/react";
import React, { useState } from "react";
import { mockIsIntersecting } from "react-intersection-observer/test-utils";
import { describe, expect, it } from "vitest";

import { Scrollama, Step } from "../src";

const stepStyle = {
  border: "5px solid #cccccc",
  height: "300px",
  width: "500px",
  margin: "100px",
};

const HookComponent = () => {
  const [currentStep, setCurrentStep] = useState<number>(0);
  return (
    <div data-testid="wrapper" data-step={currentStep}>
      <Scrollama<number> debug onStepEnter={({ data }) => setCurrentStep(data)}>
        {[1, 2, 3].map((stepNumber) => (
          <Step key={stepNumber} data={stepNumber}>
            <div style={stepStyle} data-testid={`step-${stepNumber}`}>
              Step {stepNumber}
            </div>
          </Step>
        ))}
      </Scrollama>
    </div>
  );
};

describe("Scrollama", () => {
  it("should render correctly", async () => {
    render(
      <>
        <div style={{ height: window.innerHeight }} />
        <HookComponent />
        <div style={{ height: window.innerHeight }} />
      </>,
    );
    const wrapper = screen.getByTestId("wrapper");

    // Should not be updated until intersection observer triggers
    expect(wrapper).toHaveAttribute("data-step", "0");

    const step1 = screen.getByTestId("step-1");
    const step2 = screen.getByTestId("step-2");
    const step3 = screen.getByTestId("step-3");

    act(() => {
      mockIsIntersecting(step1, true);
    });
    expect(wrapper).toHaveAttribute("data-step", "1");
    act(() => {
      mockIsIntersecting(step1, false);
    });
    expect(wrapper).toHaveAttribute("data-step", "1");

    act(() => {
      mockIsIntersecting(step2, true);
    });
    expect(wrapper).toHaveAttribute("data-step", "2");
    act(() => {
      mockIsIntersecting(step2, false);
    });
    expect(wrapper).toHaveAttribute("data-step", "2");

    act(() => {
      mockIsIntersecting(step3, true);
    });
    expect(wrapper).toHaveAttribute("data-step", "3");
    act(() => {
      mockIsIntersecting(step3, false);
    });
    expect(wrapper).toHaveAttribute("data-step", "3");

    // use screen.debug() to see the DOM
    // screen.debug();
  });
});
```

**Comment by thebuilder** [comment]
@condorheroblog you are right. I forgot to handle global correctly in the last refactor. It's tricky, since there isn't a reliable way to determine if the intersectionObserver is mocked.

**Comment by thebuilder** [comment]
@condorheroblog should be fixed #720 released in v9.16.0

**Comment by condorheroblog** [comment]
Everything is working fine, thanks👍
